lib/fieldy/writer.rb

```diff
@@ -4,38 +4,46 @@ module Writer
 
     def self.included(base)
       base.extend(WriterMethods)
+      base.send :include, InstanceMethods
+    end
+
+    module InstanceMethods
+
+      def to_s
+        data = self.class.fields
+                         .map        { |x| x[:key] }
+                         .select     { |x| x }
+                         .reduce({}) { |x, k| x.merge! k => self.send(k) }
+        self.class.write data
+      end
+
     end
-      
 
     module WriterMethods
-      
+
       def write(hash)
-        res = []
-        pack_str = ""
-        @fields.each do |h|
-          h.each do |k,v|
-            hash.each do |x,y|
-              unless k == :null
-                res << y
-              end
-            end
-          pack_str << "#{v[:type]}#{v[:length]}"
-          end
-        end
-        res.pack(pack_str)
+        values  = fields.map { |x| x[:key] }.map { |x| hash[x] || '' }
+        pack    = fields.map { |x| "#{x[:type]}#{x[:length]}" }.join ''
+
+        values.pack pack
       end
-      
-      def field(sym, length, type = "A")
+
+      def fields
         @fields ||= []
-        @fields << { sym => { :length => length, :type => type }}
       end
-      
+
+      def field(sym, length, type = "A")
```

Comment:
Plus, I don't know how you feel about this... but I was considering dropping the ```type``` argument for a more generic ```options``` hash.  I can see some possibilities for making field-level changes.  Restricting the third argument to a property that I bet will rarely change can make extending this library harder.